miden-lib/asm/kernels/transaction/api.masm

```diff
@@ -664,15 +748,50 @@ end
 #! Returns the serial number of the note currently being processed.
 #! Panics if no note is not being processed.
 #!
-#! Inputs: []
+#! Inputs: [KERNEL_PROCEDURE_HASH]
 #! Outputs: [SERIAL_NUMBER]
 #!
 #! - SERIAL_NUMBER is the serial number of the note currently being processed.
 export.get_note_serial_number
+    # drop the procedure's hash
+    dropw
+    
     exec.note::get_serial_number
     # => [SERIAL_NUMBER]
 
     # drop the padding
     swapw dropw
     # => [SERIAL_NUMBER]
 end
+
+#! Executes a kernel procedure specified by its offset.
+#!
+#! Inputs: [procedure_offset, procedure_inputs, PAD(*)]
+#! Outputs: [procedure_outputs, PAD(*)]
+#!
+#! Where:
+#! - procedure_offset is an offset of the kernel procedure, specified in the 
+#!   `miden/kernel_invocation.masm` file.
+#! - procedure_inputs are inputs of the procedure to be executed, which is specified by the 
+#!   procedure_offset. Note that the length of this inputs cannot exceed 12 elements, since the 
+#!   first word on the stack will be occupied by the procedure hash. 
+#! - procedure_outputs are the outputs of the procedure to be executed. 
+export.exec_kernel_proc
+    # compute the memory pointer at which desired procedure is stored
+    exec.memory::get_kernel_procedures_ptr add
```

Comment:
We should probably make sure that `procedure_offset` is within expected bounds - otherwise, a malicious program may execute some procedure which is actually not a kernel procedure. We can probably put a fairly low limit on the number of allowed kernel procedures - e.g., 128 or 256.

Reply:
I think to be able to properly check whether the offset is within bounds we should store the number of the kernel procedures in the selected kernel and compare the offset with it. Is it possible otherwise to execute the malicious procedure, for example, with the offset 30, which won't be a kernel procedure but will pass this assert?